crates/sui-graphql-rpc/src/server/version.rs

```diff
@@ -0,0 +1,257 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use axum::{
+    http::{HeaderMap, HeaderValue, Request, StatusCode},
+    middleware::Next,
+    response::{IntoResponse, Response},
+};
+
+use crate::error::{code, graphql_error};
+
+const VERSION_HEADER: &str = "X-Sui-RPC-Version";
+
+const RPC_VERSION_FULL: &str = env!("CARGO_PKG_VERSION");
+const RPC_VERSION_YEAR: &str = env!("CARGO_PKG_VERSION_MAJOR");
+const RPC_VERSION_MONTH: &str = env!("CARGO_PKG_VERSION_MINOR");
+
+/// Middleware to check for the existence of a version constraint in the request header, and confirm
+/// that this instance of the RPC matches that version constraint.  Each RPC instance only supports
+/// one version of the RPC software, and it is the responsibility of the load balancer to make sure
+/// version constraints are met.
+pub(crate) async fn check_version_middleware<B>(
+    headers: HeaderMap,
+    request: Request<B>,
+    next: Next<B>,
+) -> Response {
+    if let Some(req_version) = headers.get(VERSION_HEADER) {
+        let Ok(req_version) = req_version.to_str() else {
+            return (
+                StatusCode::BAD_REQUEST,
+                graphql_error(
+                    code::BAD_REQUEST,
+                    format!("Failed to parse {VERSION_HEADER}: Not an ASCII string."),
+                ),
+            ).into_response();
+        };
+
+        let Some((year, month)) = parse_version(req_version) else {
+            return (
+                StatusCode::BAD_REQUEST,
+                graphql_error(
+                    code::BAD_REQUEST,
+                    format!(
+                        "Failed to parse {VERSION_HEADER}: '{req_version}' not a valid \
+                         <YEAR>.<MONTH> version.",
+                    ),
+                ),
+            ).into_response();
+        };
+
+        if year != RPC_VERSION_YEAR || month != RPC_VERSION_MONTH {
+            return (
+                StatusCode::MISDIRECTED_REQUEST,
+                graphql_error(
+                    code::INTERNAL_SERVER_ERROR,
+                    format!("Version '{req_version}' not supported."),
+                ),
+            )
+                .into_response();
+        }
+    };
+
+    next.run(request).await
+}
+
+/// Mark every outgoing response with a header indicating the precise version of the RPC that was
+/// used (including the patch version).
+pub(crate) async fn set_version_middleware<B>(request: Request<B>, next: Next<B>) -> Response {
+    let mut response = next.run(request).await;
+    let headers = response.headers_mut();
+    headers.insert(VERSION_HEADER, HeaderValue::from_static(RPC_VERSION_FULL));
+    response
+}
+
+/// Split a `version` string into two parts (year and month) separated by a ".".
+///
+/// Confirms that the version specifier contains exactly two components, and that both
+/// components are entirely comprised of digits.
+fn parse_version(version: &str) -> Option<(&str, &str)> {
```

Comment:
I went back and forth on this, especially because I just introduced the `SuiRpcVersion` TypedHeader, and it's possible to implement validation during header parsing.  I ended up not following through with this comment for a couple of reasons:

axum is responsible for creating the `SuiRpcVersion` instance that gets fed to the middleware.  If I fold the logic for parsing the header into there, it will also be responsible for outputting the error if that parsing fails.  AFAICT, there isn't a way to customise the error response if I set things up this way, so it becomes impossible to provide more context on what was wrong with the request, so `SuiRpcVersion` just extracts the raw header, and we need to parse it in the middleware body and construct an error in the GraphQL style if that goes wrong.

I could then have introduced another type for a parsed header, but chose not to because it's only used in one place internal to this module, the type would require lifetime annotations because it holds references, and could easily be confused with the typed header.